diagnostics-app/src/app/core/services/events.service.ts

```diff
@@ -0,0 +1,124 @@
+// Copyright 2023 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/**
+ * @fileoverview Service for starting to capture events and stopping
+ * capturing events from Chrome extension.
+ */
+
+import { Injectable } from '@angular/core';
+import { Subject } from 'rxjs';
+
+import { environment } from 'src/environments/environment';
+import {
+  EventCategory,
+  EventsInfo,
+} from '@common/dpsl'
+import {
+  EventsAction,
+  EventMessage,
+  EventsRequest,
+  EventsResponse,
+  Request,
+  RequestType,
+  Response,
+  PortName,
+} from '@common/message';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class EventsService {
+  private extensionId!: string;
+  private subjects = new Map<string, Subject<EventsInfo>>;
+  private port?;
+
+  private constructEventsRequest: (
+    payload: EventsRequest
+  ) => Request = payload => {
+    return { type: RequestType.EVENTS, events: payload };
+  };
+
+  private sendRequest: (request: Request) => Promise<EventsResponse> = (
+    request: Request
+  ) => {
+    return new Promise((resolve, reject) => {
+      try {
+        //@ts-ignore
+        window.chrome.runtime.sendMessage(
+          this.extensionId,
+          request,
+          (response: Response) => {
+            if (!response.success) {
+              reject(response.error);
+            } else if (!response.events) {
+              throw 'Invalid response';
+            } else {
+              resolve(response.events);
+            }
+          }
+        );
+      } catch (err) {
+        reject(err);
+      }
+    });
+  };
+
+  constructor() {
+    this.extensionId = environment.extensionId;
+
+    try {
+      //@ts-ignore
+      this.port = window.chrome.runtime.connect(
+        environment.extensionId,
+        { name: PortName.EVENTS_PORT }
+      );
+
+      this.port.onMessage.addListener((msg: EventMessage) => {
+        if (this.subjects.has(msg.type)) {
+          this.subjects.get(msg.type)!.next(msg.info);
+        } else {
+          console.error('Error when mapping events type to get subject.');
+        }
+      });
+
+    } catch (err) {
+      console.error('Error when establishing connection and add listener with extension.');
+    }
+  }
+
+  getSubject: (category: EventCategory) => Subject<EventsInfo> = (category) => {
```

Comment:
What is the flow here for `getSubject` and `isEventSupported`?

Ideally, we should never allow a subject to register to an event that is not supported. 

One possible way is the let `getSubject` check for event supportability before setting the subject. Let me know your thoughts on this.

Reply:
In the current design, the events dashboard component will call the <code>isEventSupported</code> to check which event categories can be displayed in the dashboard. And the <code>getSubject</code> will be called in the card components when they are initializing (in the <code>ngOnInit</code> block).

Hence, there won't be a case that <code>getSubject</code> is called for an unsupported event category. And since the event service creates the subject for the event category only when <code>getSubject</code> is called, there won't be any subject that registers to an unsupported event.

Reply:
I see.

I'm still slightly uncomfortable with the current design that anyone can register a subject, and will succeed even if the routine is non-existent.

How about we either 
1. Check `IsEventSupported` when trying to register for the subject
2. maintain a set called `SupportedEvent`. In the initial call of `IsEventSupported`, add the event type into the set. When registering for `GetSubject`, check whether the requested type is already in SupportedEvent.

Reply:
+1 to check event supported in `getSubject`.

Although we call `getSubject` for supported event only, it will be more robust to ensure the event is supported before getting subject. It can help us avoid unexpected bugs on the UI side in the future.

Reply:
In a future CL, I've added the ability to cache the results from getEventSupportability. This is used primarily to ensure that all supported events will be loaded in due time (navigating between pages will cause a complete refresh of supported events, which stabilize after 1~2 seconds).

I will add the check for getSubject in that future CL, since without caching, calling IsEventSupported in getSubject will incur significant performance cost.